tempesta_fw/hpack.c

```diff
@@ -0,0 +1,3439 @@
+			if (likely(shift <= hp->curr + HT_NBITS)) {
+				sym = (char)ht_decode[offset + i].offset;
+				if (tfw_hpack_huffman_write(sym, req))
+					return T_DROP;
+				hp->curr -= shift;
+				return huffman_decode_tail(hp, req, 0);
+			}
+		} else {
+			/*
+			 * Condition here equivalent to the
+			 * "-shift <= hp->curr + HT_NBITS", but
+			 * working faster.
+			 */
+			if (shift >= -HT_NBITS - hp->curr) {
+				if (ht_decode[offset + i].offset == 0) {
+					return T_DROP;
+				}
+			}
+		}
+	}
+	return T_DROP;
+}
+
+static int
+tfw_huffman_decode(TfwHPack *__restrict hp, TfwHttpReq *__restrict req,
+		   const unsigned char *__restrict src, unsigned long n)
+{
+	unsigned int offset;
+
+	WARN_ON_ONCE(n > hp->length);
+	if (!n)
+		return T_OK;
+
+	SET_NEXT();
+	for (;;) {
+		offset = 0;
+		for (;;) {
+			int shift;
+			unsigned int i;
+
+			if (hp->curr <= 0) {
+				if (likely(n)) {
+					SET_NEXT();
+				} else if (hp->length) {
+					return T_POSTPONE;
+				} else {
+					/*
+					 * Last full prefix also processed here
+					 * (see hp->curr <= 0 above).
+					 */
+					return huffman_decode_tail(hp, req,
+								   offset);
+				}
+			}
+			i = (hp->hctx >> hp->curr) & HT_NMASK;
+			shift = ht_decode[offset + i].shift;
+			offset = ht_decode[offset + i].offset;
+			T_DBG3("%s: shift, hp->curr=%d, hp->hctx=%hx,"
+			       " hp->length=%lu, n=%lu, i=%u, shift=%d,"
+			       " offset=%u, offset=%c\n", __func__, hp->curr,
+			       hp->hctx, hp->length, n, i, shift, offset,
+			       (char)offset);
+			if (shift >= 0) {
+				if (tfw_hpack_huffman_write((char)offset, req))
+					return T_DROP;
+				hp->curr -= shift;
+				offset = 0;
+			} else {
+				hp->curr += shift;
+				if (offset >= HT_SMALL) {
+					break;
+				}
+				if (unlikely(offset == 0)) {
+					goto end;
+				}
+			}
+		}
+		hp->curr += HT_NBITS - HT_MBITS;
+		/*
+		 * With various optimization options, the anonymous block here
+		 * leads to the generation of more efficient code.
+		 */
+		{
+			int shift;
+			unsigned int i;
+
+			if (hp->curr < 0) {
+				if (likely(n)) {
+					SET_NEXT();
+				} else if (hp->length) {
+					return T_POSTPONE;
+				} else {
+					return huffman_decode_tail_s(hp, req,
+								     offset);
+				}
+			}
+			i = (hp->hctx >> hp->curr) & HT_MMASK;
+			shift = ht_decode[offset + i].shift;
+			offset = ht_decode[offset + i].offset;
+			T_DBG3("%s: short shift, hp->curr=%d, hp->hctx=%hx,"
+			       " hp->length=%lu, n=%lu, i=%u, shift=%d,"
+			       " offset=%u, offset=%c\n", __func__, hp->curr,
+			       hp->hctx, hp->length, n, i, shift, offset,
+			       (char)offset);
+			if (likely(shift >= 0)) {
+				if (tfw_hpack_huffman_write((char)offset, req))
+					return T_DROP;
+				hp->curr -= shift;
+			} else {
+				break;
+			}
+		}
+	}
+end:
+	return T_DROP;
+}
+
+static inline TfwHPackStr *
+tfw_hpack_create_str(unsigned long len)
+{
+	TfwHPackStr *str;
+
+	if (!(str = kmalloc(sizeof(TfwHPackStr) + len, GFP_ATOMIC)))
```

Comment:
Yeah - I forgot to replace these `kmalloc` calls with variable-sized entries in ring buffer. Also during recent HTTP/2 parser implementation I have stumbled upon another issue regarding the headers storage in decoder index table: to preserve structure of already parsed headers in order to avoid redundant parsing of indexed (and consequently, already parsed message) - we need to store TfwStr descriptor in ring buffer for name and value of the header. Since I've began to make necessary corrections in the last PR - thus, I will make all these changes in decoder ring buffer in context of last PR.